sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ClassAndMethodName.cs

```diff
@@ -0,0 +1,341 @@
+
+            var parts = SplitToParts(identifierName).ToList();
+            for (int i = 0; i < parts.Count; i++)
+            {
+                string part = parts[i];
+                if (part.Length == 1 && part[0] == '_' && !areUnderscoresAllowed)
+                {
+                    continue;
+                }
+
+                var ideal = i == 0
+                    ? HandleFirstPartOfTypeName(part, requireInitialI, allowInitialI, 1)
+                    : SuggestFixedCaseName(part, 1);
+
+                var acceptable = i == 0
+                    ? HandleFirstPartOfTypeName(part, requireInitialI, allowInitialI, 2)
+                    : SuggestFixedCaseName(part, 2);
+
+                idealNameVariant.Append(ideal);
+                acceptableNameVariant.Append(acceptable);
+            }
+
+            suggestion = SuggestCapitalLetterAfterNonLetter(idealNameVariant);
+            var acceptableSuggestion = SuggestCapitalLetterAfterNonLetter(acceptableNameVariant);
+
+            return acceptableSuggestion == identifierName ||
+                   suggestion == identifierName;
+        }
+
+        private static string HandleFirstPartOfTypeName(string input,
+            bool requireInitialI, bool allowInitialI, int maxUppercase)
+        {
+            bool startsWithI = input[0] == 'I';
+
+            if (requireInitialI)
+            {
+                string prefix = startsWithI ? string.Empty : "I";
+                return prefix + SuggestFixedCaseName(FirstCharToUpper(input), maxUppercase + 1);
+            }
+
+            string suggestionToProcess;
+            if (input.Length == 1 &&
+                !allowInitialI &&
+                startsWithI &&
+                IsCharUpper(input, 0) &&
+                !IsCharUpper(input, 1))
+            {
+                suggestionToProcess = FirstCharToUpper(input.Substring(1));
+            }
+            else
+            {
+                suggestionToProcess = FirstCharToUpper(input);
+            }
+
+            return SuggestFixedCaseName(suggestionToProcess, maxUppercase);
+        }
+
+        private static string SuggestCapitalLetterAfterNonLetter(StringBuilder suggestion)
+        {
+            for (int i = 1; i < suggestion.Length; i++)
+            {
+                if (!char.IsLetter(suggestion[i - 1]) &&
+                    char.IsLower(suggestion[i]))
+                {
+                    suggestion[i] = char.ToUpperInvariant(suggestion[i]);
+                }
+            }
+            return suggestion.ToString();
+        }
+
+        private static string SuggestFixedCaseName(string input, int maxUppercaseCount)
+        {
+            var upper = input.Take(maxUppercaseCount);
+            var lower = input.Skip(maxUppercaseCount).Select(char.ToLowerInvariant);
+
+            return new string(upper.Concat(lower).ToArray());
+        }
+
+        private static IEnumerable<string> SplitToParts(string name)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3776)